psa-crypto/src/types/mod.rs

```diff
@@ -2,7 +2,44 @@
 // SPDX-License-Identifier: Apache-2.0
 
 //! # PSA Types
-
 pub mod algorithm;
 pub mod key;
 pub mod status;
+
+use zeroize::Zeroize;
+
+macro_rules! zeroize_on_drop {
+    ($type:ty) => {
+        impl Drop for $type {
+            fn drop(&mut self) {
+                self.zeroize();
+            }
+        }
+    };
+}
+
+zeroize_on_drop!(algorithm::Algorithm);
+zeroize_on_drop!(algorithm::Hash);
+zeroize_on_drop!(algorithm::FullLengthMac);
+zeroize_on_drop!(algorithm::Mac);
+zeroize_on_drop!(algorithm::Cipher);
+zeroize_on_drop!(algorithm::AeadWithDefaultLengthTag);
+zeroize_on_drop!(algorithm::Aead);
+zeroize_on_drop!(algorithm::SignHash);
+zeroize_on_drop!(algorithm::AsymmetricSignature);
+zeroize_on_drop!(algorithm::AsymmetricEncryption);
+zeroize_on_drop!(algorithm::RawKeyAgreement);
+zeroize_on_drop!(algorithm::KeyAgreement);
+zeroize_on_drop!(algorithm::KeyDerivation);
+
+zeroize_on_drop!(key::DhFamily);
+zeroize_on_drop!(key::EccFamily);
+zeroize_on_drop!(key::Type);
+zeroize_on_drop!(key::Lifetime);
+zeroize_on_drop!(key::Policy);
+zeroize_on_drop!(key::Id);
+zeroize_on_drop!(key::Attributes);
+zeroize_on_drop!(key::UsageFlags);
+
+zeroize_on_drop!(status::Status);
+zeroize_on_drop!(status::Error);
```

Comment:
Do all those types always carry confidential information? Do they all contain valuable data for an attacker?

Reply:
Don't think so, but I went for the full interpretation of the spec where all temporary values must be cleaned up.

Reply:
Plus if you think about a full description of a key, you need to clear all the params associated with it - I chose the easy way out 😅 otherwise I'd have to implement `Zeroize` by hand on the types that don't have all inner values zeroize'able

Reply:
Also, if there's an efficiency problem here I can just make the `zeroize_on_drop` conditional on a feature.

Reply:
Linking with [this issue](https://github.com/parallaxsecond/parsec/issues/122) for reference. I am not sure if the specification is that strict [on the section about memory cleanup](https://armmbed.github.io/mbed-crypto/html/overview/implementation.html#memory-cleanup).

> All temporary data used during the execution of a function, such as stack buffers, must be wiped before the function returns.

"stack buffer" given as an example might imply that this rule applies only to buffers.

> Implementations must wipe all sensitive data from memory when it is no longer used. It is recommended that they wipe this sensitive data as soon as possible.

Question is to know if those are parts of the "sensitive data". I would be careful about implementing such an impactful change if it is not needed, or negligible next to the memory cleaning of buffers. I will ask about that point.

Reply:
Sure, tbf it's probably a good idea to not have the memory wiping by default - I think having the `Zeroize` derived on the types makes sense, though

Reply:
We can also run some tests and see the exact performance figure with and without it